pkg/spanconfig/spanconfigsqlwatcher/buffer.go

```diff
@@ -119,27 +120,84 @@ func (b *buffer) flushEvents(
+) (sqlUpdates []spanconfig.SQLUpdate, _ hlc.Timestamp, _ error) {
 	events, combinedFrontierTS := b.flushEvents(ctx)
+	// The events slice can contain a SQLUpdate that applies to a descriptor, or a
+	// SQLUpdate that applies to a protected timestamp record. We bubble up the
+	// protected timestamp SQLUpdates to the end of the slice, so that we can
+	// perform the subsequent deduplication of SQLUpdates. We do this instead of
+	// pre-processing events to separate the two kinds of updates, to save on the
+	// allocation of additional `events` slices.
 	sort.Slice(events, func(i, j int) bool {
 		ei, ej := events[i].(event), events[j].(event)
-		if ei.update.ID == ej.update.ID {
-			return ei.timestamp.Less(ej.timestamp)
+		if ei.update.IsDescriptorUpdate() && ej.update.IsDescriptorUpdate() {
+			descUpdatei := ei.update.GetDescriptorUpdate()
+			descUpdatej := ej.update.GetDescriptorUpdate()
+			if descUpdatei.ID == descUpdatej.ID {
+				return ei.timestamp.Less(ej.timestamp)
+			}
+			return descUpdatei.ID < descUpdatej.ID
 		}
-		return ei.update.ID < ej.update.ID
+
+		// If the LHS is a descriptor update, sort it before the RHS which is a
+		// protected timestamp update.
+		if ei.update.IsDescriptorUpdate() {
+			return true
+		}
+
+		// If the RHS is a descriptor update, sort it before the LHS which is a
+		// protected timestamp update.
+		if ej.update.IsDescriptorUpdate() {
+			return false
+		}
+
+		return ei.timestamp.Less(ej.timestamp)
+	})
+
+	// Find the index before which all events are SQLUpdates on descriptors.
+	descriptorUpdatesIdx := sort.Search(len(events), func(i int) bool {
+		update := events[i].(event).update
+		return update.IsProtectedTimestampUpdate()
 	})
-	for i, ev := range events {
-		if i == 0 || events[i-1].(event).update.ID != ev.(event).update.ID {
-			updates = append(updates, ev.(event).update)
+
+	// Deduplicate the SQLUpdates emitted by the buffer that apply to descriptors.
+	for i, ev := range events[:descriptorUpdatesIdx] {
+		update := ev.(event).update
+		descriptorUpdate := update.GetDescriptorUpdate()
+		if i == 0 {
+			sqlUpdates = append(sqlUpdates, update)
 			continue
 		}
-		descType, err := combine(updates[len(updates)-1].DescriptorType, ev.(event).update.DescriptorType)
+
+		prevUpdate := events[i-1].(event).update
+		if prevUpdate.GetDescriptorUpdate().ID != descriptorUpdate.ID {
+			sqlUpdates = append(sqlUpdates, update)
+			continue
+		}
+
+		prevDescriptorSQLUpdate := sqlUpdates[len(sqlUpdates)-1].GetDescriptorUpdate()
+		descType, err := combine(prevDescriptorSQLUpdate.DescriptorType,
+			descriptorUpdate.DescriptorType)
 		if err != nil {
 			return nil, hlc.Timestamp{}, err
 		}
-		updates[len(updates)-1].DescriptorType = descType
+		sqlUpdates[len(sqlUpdates)-1] = spanconfig.MakeDescriptorSQLUpdate(
+			prevDescriptorSQLUpdate.ID, descType)
+	}
+
+	// Deduplicate the SQLUpdates emitted by the buffer that apply to protected
+	// timestamps.
+	ptsUpdates := make(map[string]struct{})
```

Comment:
Do we need to use a map here if the pts updates are sorted stably? Shouldn't we be able to compare against the last entry in the slice like we do above?

Reply:
As mentioned above, I'm unsure how to make sorting between two TenantTargets stable. I think it is simpler to just use their string protos as a key as it is today?